readme.md
# PyCraft Server #
Pycraft Server is a collection of tools that can be used while running a server. Basic usage allows running a minecraft server with arguments more easily, run backups and automatically shut down the server when no players are online for some time.

PyCraft Server is a bundle of python scripts written by [AgentM](https://www.youtube.com/c/AgentMOfficial)  
Special thanks to Dinnerbone for [mcstatus](https://github.com/Dinnerbone/mcstatus)

**Documentation version: 1.1**  
**Script version: `<can be seen when running server.py>`**  
*Make sure these match before contacting me.*

## Table of Contents ##
 1. [Installation](#install)
 2. [Basic Usage](#basic)
   - Setting up a PyCraft server
     - PyCraft file structure
     - PyCraft server config file
   - Running a PyCraft server
     - PyCraft command line
   - Using built-in commands
     - help
     - stop | quit | exit
     - modules
 3. [Built-in Modules](#modules)
   - auto-shutdown | as
   - backup
   - status
   - notify
 4. [Advanced Usage](#advanced)
   - Creating your own modules
     - Bare bones
     - Running server commands
     - Server Config
     - Events
 5. [Extended Functionality](#extended)
   - Features
     - Auto updater
   - Security Patches

<a name="install">

## 1. Installation ##
</a>

You need Python 3.X to run the scripts.  
Also, install mcstatus from: [https://github.com/Dinnerbone/mcstatus](https://github.com/Dinnerbone/mcstatus)

Then, installation is as simple as:
1. Extracting the contents of the zip to some safe folder (I like to put it under C:/MCServers or something).
2. Putting a server.jar in a ServerName folder in servers.
3. (Running the server to generate a server.properties and eula, if you haven't already)
4. Creating a configuration for this server by running `python pycraft_config.py`
5. Running `python pycraft.py ServerName`.
6. (Optionally) Run `python pycraft_updater.py` in conjunction with `auto-update` set to `true` in `config.json` to update all your servers.

If you have any trouble, check my YouTube channel for the tutorial: [https://www.youtube.com/c/AgentMOfficial](https://www.youtube.com/c/AgentMOfficial)  
You can contact me in the comments for any further questions.

<a name="basic">

## 2. Basic Usage ##
</a>

### Setting up a PyCraft server ###

#### PyCraft file structure
The PyCraft server will need a couple files and folders to work.

```
- (root)
  - [modules]
    - auto_shutdown.py
    - backup.py
    - status.py
    - ...
  - [servers]
    - [SomeServer]
      - [backups]
        - [auto]
          - world1_2020-12-01_23-55-00.zip
        - world1_2020-12-01-23-54-00.zip
      - [logs]
      - [worlds]
        - [world1]
          - level.dat
          - ...
      - eula.txt
      - server.jar
      - server.properties
      - server-icon.png
      - ...
  - config.json
  - pycraft.py
  - pycraft_config.py
  - pycraft_module.py
  - pycraft_updater.py
  - pycraft_utils.py
  - readme.md
```

The `...` represent other files and folders, for example, those created by the server.jar or custom modules.

For every server folder ("SomeServer" in the example above), there must be an entry in the config.json file specifying how to run it. The structure of such a config.json file is given below.

#### PyCraft server config file ####

A config.json file contains the configuration information for the PyCraft and its modules on how to run the servers and with which parameters.

You can run `python pycraft_config.py` from the command line to create and edit such config file.

The basic barebones config file looks like this:

``` json
{
    "jvm-args": [],
    "hide-gui": true,
    "upgrade-all-chunks-on-version-mismatch": false,
    "module-data": {
      "module_<module_name>": {
        "module_config_setting": "Any"
      }
    },
    "server-list": [
        {
            "name": "ServerName",
            "universe": "worlds",
            "world": "test_world",
            "version": "custom",
            "port": 25565,
            "description": "Test server for the development of PyCraft.",
            "auto-update": false,
            "auto-restart": false,
            "read-only": false,
            "initialize": [
                "modules list",
                "/say This message will appear when the server is ready."
            ]
        }
    ]
}
```

These config settings apply to every server configuration that is run. For any specific server configuration settings, apply them in "server-list" under the server for which you want to apply that config.

- `jvm-args` (list\<str\>): A list of str arguments to pass to the java virtual machine (to increase RAM for example)
- `hide-gui` (bool): Hide the server console window from popping up.
- `upgrade-all-chunks-on-version-mismatch` (bool): If the server should upgrade/optimize chunks when it has recently been updated to a different version.
- `module-data` (dict): Any custom configuration settings used by modules. The convention is to use `module_<module_name>` for the key to properly namespace settings. `shared` could be used for any config settings shared between modules.
<!--"upgrade-all-chunks-on-version-mismatch" will probably be moved to server specific config-->


- `server-list` (list\<dict\>): A list of all server configurations. Each server must have the following keys:
  - `name` (str): The exact same name as the server folder (server.jar is in this folder)
  - `universe` (str): The name of the universe (location of all server worlds).
  - `world` (str): The name of the world.
  - `version` (str): The type of version the server runs on (for auto updates), can be "release" or "snapshot", "custom" if you don't want to use auto-updates, or if using modded versions.
  - `port` (int\<0-65536\>): The server port to use. Note that the server will run on port 25565 if this isn't specified. It will NOT use the port specified in server.properties, this is completely ignored.
  - `description` (str): Human readable description for what the server is for.
  - `auto-update`<a name="autoupdates"> </a>(bool): For 'release' or 'snapshot' versions, will automatically check for updates and apply them to the server when the server is booted up. **`Note`**: This setting is not checked by `pycraft.py` but only by the `pycraft_updater.py`. You may run the updater directly before the server each time to make use of this feature effectively.
  - `auto-restart` (bool): *`Not yet implemented`*; If the server should automatically restart when it crashes (not when it gracefully closes)
  - `read-only` (bool): *`Not yet implemented`*; If the map loaded should be saved to. If this is turned on, the server will make a temporary copy of the world that is selected. `<WORLDNAME_pycraft_copy>` (overriding the previous one). This is useful when you want to run minigames or custom maps that need to be in pristine condition when you first start it. The server will not reset the map when it closed due to a crash, this to preserve the state. You can also just save the copy under a different name to keep progress, but then why are you using read-only anyways?
  - `initialize` (list\<str\>): A list of commands ran at server startup. Commands that start with a `/` are server commands such as `/say`, `/give`, etc. Other commands are module commands such as `modules list` (to list all active modules) (for example setting automatic shutdown and backups, or to send a nice log message, or other stuff)
  - `__comment_<anything>`: Comments, useful for reading configurations as a human, but it doesn't really do anything for us computers.

### Running a PyCraft server ###

#### PyCraft command line ####
You can start the PyCraft server from the command line using:
```bat
python pycraft.py [YourServerName]
```
If you want you can simply put the line above in a windows batch file (server.bat) and run it by right clicking. (For linux/mac you can use a .sh file)

### Using built-in commands ###
The difference between this section and the next (modules), is that these commands are hardcoded, because they tie in with the core functionality and should not be accidentally removed from the 'modules' folder for example.

#### help ####
Use this command to view help of all available commands and modules.

- `help`: To get an overview of all commands
- `help <MODULE> [subcommands]`: Show the help of MODULE, with optional 'subcommands' for more detailed info.

#### stop | quit | exit ####
You can use any of those 3 commands to gracefully stop the server.

*This command takes no arguments.*

#### modules ####
Use this command to manage modules.

- `modules list`: List all active modules.
- `modules reload`: Reload all modules (any found in 'modules' folder will be available for usage)

<a name="modules">

## 3. Built-in Modules ##
</a>

The built-in modules are not strictly required, but are vital that they are included with every shipment of PyCraft.

### auto-shutdown | as ###
This command can be used to manage automatic shutdown of the server.
**It will also shut down the computer if HARD is specified.**

- `as`: Shows the usage of this command.
- `as idle <TIME> [HARD]`: Schedules a shutdown when no players are online.
- `as schedule <TIME> [HARD`: Schedules a shutdown right now. (reminds players 1m and 10s before shutdown)
- `as cancel`: Cancels any schedule or idle timers.

Example: `as idle 10m HARD` Will shutdown the server AND computer after nobody has been online for 10 minutes. The 10m timer is reset when a player joins. This can also be canceled completely using `as cancel`.

The shutdown thread will query the server for the amount of players online, for short schedules, there's at least 10 seconds between each request to the server, for longer schedules, there's up to 120 seconds between each request, to reduce overloading the server. This means that after a player has left, it may take up to 120 seconds longer for the script to detect the player count on the server. A 30 minute idle schedule can therefore take up to 32 minutes. However a 60 seconds idle timer may take up to 70 seconds, depending on the polling time. For your interest:

``` python
poll_delay = max(10, min(t // 10, 120))
```

### backup ###
This command handles backups. There are 2 types of backups, manual backups and automatic backups. They are stored separately.

Manual backups are stored under `<SERVER FOLDER>/backups`.  
Automatic backups are stored under `<SERVER FOLDER>/backups/auto`.

- `backup now [END]`: Creates a manual backup right now. (Specify `END` to close the server after the backup finishes)
- `backup schedule <TIME> [AMOUNT]`: Schedules a backup in TIME up to AMOUNT auto backups in total, after which the oldest is deleted.
- `backup off`: Turns off automatic backups.

The name of the resulting zipfile backup will be: `<universe>_<world>_<date>_<time>.zip`  
Automatic backups will have the name: `<universe>_<world>_<date>_<time>_apcbkp.zip`

**Note that making a backup could take a while and during a backup, auto-saving is turned off. The absolute bare minimum backup-time is therefore 2 minutes, 30 minutes or more is advised. However, using 7z combined with fast-backups, backing up may be faster.** <!--Currently, the zipping is not smart and will include all files, not just the new/changed/deleted files. However using 7z quick times like 15s for 2 to 3 GB can be achieved easily-->

**Also note that a backup takes quite some space, especially for large worlds. Therefore it is recommended to keep the total amount of automatic backups around 4 max. You can make as many manual backups as you'd like of course.**

*The server will NOT automatically back up any worlds when the version is updated.* <!--Make sure to make a backup before updating to a newer version automatically.-->

#### 7z assisted backups ####
Some configuration options exist to increase the efficiency of backups using a third party tool: 7-Zip. (Not affiliated)

* `7z-path`
  * The path to your 7z command line executable. (or simply leave it out if 7z is on your path)

* `prefer-7z`
  * Tries to use 7z over the standard backup method (if it can succesfully locate the executable).

* `fast-backup`
  * Requires 7z utilization. If a previous backup file can be found in the backups directory, it will be used as a basis for the new backup. This can speed up backups significantly for large worlds with only few modifications. (Basically skips: any regions that you haven't loaded/unmodified files)

An example setup in the config.json:

```json
"module-data": {
  "module_backup": {
    "prefer-7z": true,
    "fast-backup": true,
    "7z-path": "C:\\Program Files\\7-Zip\\bin\\7z.exe"
  }
}
```

### status ###
A simple wrapper for mcstatus. Displays some useful server information such as: ping, version, description, players and query. (if query is enabled)

### notify ###
A simple tool to make the console window go *BEEP* whenever certain criteria are met, such as: someone entering the server, someone leaving the server, specific chat messages, etc.

<a name="advanced">

## 4. Advanced Usage ##
</a>

### Creating your own modules ###
To start creating your own module, create a new file `<nameHere>.py` under `modules`. Then edit it with the editor of your likings. You will need to know a little bit of python and at least the bare bones (boilerplate) code given below to start.

#### Bare bones ####

For starters I recommend checking the other modules and figure out how they work first. This will give you an insight for writing your own.

The boilerplate (required structure of a module) is as follows:

``` python
from pycraft_module import PCMod

description = "Place a human friendly description here."
patterns = ['custom-command']

def get_module():
    return module

def usage(subcmd=[]):
    return "Replace this with your help message."

# Not strictly necessary, but generally a stylish way to print commands.
# Don't use raw print statements!
def pyprint(string, loglevel=1):
    get_module().pyprint(string, loglevel)

def callback(cmd, server_config, run_cmd, event_triggers):
    '''
    cmd: The remaining subcommands after the matched pattern.
    server_config: The server config specific to this server, including port, name, etc.
    run_cmd: Allows writing commands to the server such as say, stop, etc. (Without preceding '/'!)
    event_triggers: Dict of events that can be triggered (see pycraft_server for a list of them).
    '''

    ### Write your module code here.
    pass

module = PCMod(__name__, description, patterns, callback, usage)

```


The boiler plate code can be even more concise when you remove all the comments, of course.

Most of the time you may also want to use

``` python
import pycraft_utils as pu
```
This includes easy argument parsing and various neat utilities.

#### Running server commands ####

You can run commands on the server by simply calling `run_cmd(<YOUR COMMAND HERE>)`
You can run any server command, but be careful with them, things like `fill`, `kill` can do major damage to the server when used wrongly.

If you want to log something, don't use say. This will print to all players. You can just use pyprint(msg), this will only log to the pycraft server, but that's the best we can do. Or you'd need to do some fancy trickery.

#### Server Config ####

The server specific config file is passed as a dictionary. You can edit it, but it's recommended you don't, as other modules might depend on it. You can read/write custom key-value pairs to it, but the changes will only be local, as a server restart will reset them. <!--This might change in the future-->

#### Events ####

Some things can't be queried using namemc/mcstatus. I provided as best as I could, an interface between server chat and events for the major things.

`event_triggers` is passed as a dict, with keys as event names.
e.g. `{'done': event_trigger}`

The keys for events include:

- `done`: Triggers when the server is done loading and is ready to receive commands.
- `save`: Triggers when the server is saved using save-all.
- `stop`: Triggers when the server is stopped. (for shutdown hooks only, spawn a thread waiting for this event.)
- `join`: Triggers when a player joins the server.

changelog.txt
PYCRAFT CHANGELOG
- last update: 10 Dec, 2021

This changelog is an incomplete history of minor versions.
Smaller patches are directly applied to minor versions. Therefore, there may be differences between two instances of version 1.1.
The minor versioning is only updated when there's a critical update for security, older versions than the latest should probably not be used.

===================== LATEST STABLE =====================

=== 1.1 ===
 - Apply patch for CVE-2021-44228: Log4j RCE 0-day mitigation through custom log4j configurations.


===================== HISTORICAL =====================

=== 1.0 ===
 - Initial version (now insecure to use)

pycraft.py
### pycraft_server.py
## Script provided as-is by AgentM
## Handles a minecraft server.
import pycraft_utils as pu

import pycraft_module
import subprocess
import importlib
import argparse
import glob
import json
import time
import sys
import zipfile
import os
import re

from mcstatus import MinecraftServer
from jprops import Properties
from threading import Thread
from threading import Event
from threading import Lock
from threading import Condition
from queue import Queue
from os import path

# Make sure our working directory is the location of the pycraft_server.py file.
cur_dir = path.dirname(__file__)
if len(cur_dir.strip()) > 0:
	os.chdir(cur_dir)

DEBUG = False
if DEBUG:
	pycraft_module.log_flag |= 0b1111

#GLOBALS
# Version history: 1.0 is unsafe due to log4j error. Use 1.1+
pycraft_server_version = '1.1'

config_file = 'config.json'
servers_location = 'servers'
backups_location = 'backups'
modules_location = 'modules'
resources_location = 'resources'
sys.path.insert(1, modules_location) # Tell python to look in the modules folder when relaoding imports.

server_config = None
server_properties = None
queue_lock = Lock()
read_flag = False
running = False
command_providers = []
raw_imports = []
read_condition = Condition()

base_pattern = '(^\\[\\d{1,2}:\\d{1,2}:\\d{1,2}\\] \\[[a-zA-Z\\s]*?(?:|#\\d+)\\/[A-Z].*?\\]:) (%s)'
name_pattern = '[a-zA-Z0-9_]+?' # Use this when you don't use pre-/suffixes (safer)
ps_name_pattern = '[^<*].*' # Can be anything because of pre-/suffixes BUT disallows < and > usage. (more lenient, less safe)

signature_done = re.compile(base_pattern % 'Done \\([0-9.]+s\\)! For help, type "help"')
signature_save = re.compile(base_pattern % 'Saved the game')
signature_stop = re.compile(base_pattern % 'Stopping server')
signature_join = re.compile(base_pattern % ('UUID of player %s is [0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$' % name_pattern))
signature_leave = re.compile(base_pattern % ('%s lost connection: .*$' % name_pattern))
signature_chat = re.compile(base_pattern % '<.*?> .*')
signature_server_chat = re.compile(base_pattern % '[Server] .*') # Not safe. May also trigger on entities or commandblocks named 'Server' performing the /say command.
signature_emote = re.compile(base_pattern % '\\* .*')
signature_any = re.compile(base_pattern % '.*')

class EventTrigger:
	def __init__(self, signature):
		self.event = Event()
		self.signature = signature
		self.data = None
		self.match = None

event_triggers = {
	'done': EventTrigger(signature_done), # Triggers when the server is done loading and is ready to receive commands.
	'save': EventTrigger(signature_save), # Triggers when the server is saved using save-all.
	'stop': EventTrigger(signature_stop), # Triggers when the server is stopped. (for shutdown hooks only, spawn a thread waiting for this event.)
	'join': EventTrigger(signature_join), # Triggers when a player joins the server.
	'leave': EventTrigger(signature_leave), # Triggers when a player leaves the server.
	'chat': EventTrigger(signature_chat), # Triggers on any chat message (starting with <NAME>)
	'server-chat': EventTrigger(signature_server_chat), # Triggers on any chat message sent by the SERVER ONLY. (or a player named Server, be careful, don't give them '/say' access)
	'emote': EventTrigger(signature_emote), # Triggers on all emotes (lines starting with *).
	'any': EventTrigger(signature_any), # Triggers on anything, useful for partially regexxing.
}

safety = ""
severities = ['DEBUG', 'INFO', 'WARN', 'ERROR']
def pyprint(string, loglevel=1):
	print(f"[{safety}PyCraft/%s] %s" % (severities[loglevel], string))

def arguments():
	parser = argparse.ArgumentParser(description="Run a minecraft server with backups, shut-down hooks and possibly scripts.", epilog="Priority Order: Arguments, config.json, server.properties, DEFAULT")

	# Required
	parser.add_argument(dest="server_name", help="Starts the server with the specified name, matching case. If not specified, an interactive console allows you to select one.", nargs='?', default=None)
	
	# Optional
	parser.add_argument("-j", "--jvm-args", default=[], nargs="+", dest="jvm_arguments", help="Run the server with more RAM or other jvm arguments (don't include the leading dashes).")
	parser.add_argument("-u", "--universe", default=None, dest="universe", help='Select a folder as the save location of world folders for the server. (see Priority Order)')
	parser.add_argument("-w", "--world", default=None, dest="world", help="Select a folder as the world folder to load for the server. (see Priority Order)")
	return parser.parse_args()

def read_config():
	try:
		with open(config_file, 'r') as f:
			return json.load(f)
	except json.decoder.JSONDecodeError as e:
		pyprint("Error in Config file: %s" % str(e), 3)
		sys.exit(1)

def find_server_config(server_name, dict_list):
	for d in dict_list:
		if d['name'] == server_name:
			return d
	raise Exception('[Config] Configuration for %s was not found!' % server_name)

def get_server_properties(server_jar_location):
	p = Properties()
	with open(path.join(server_jar_location, 'server.properties'), 'r') as f:
		p.load(f)
	return p

def server_args(universe, world, nogui, forceupgrade, port):
	args = []
	if nogui: args.append('--nogui')
	if forceupgrade: args.append('--forceUpgrade')
	if universe != '.': args.extend(['--universe', universe])
	args.extend(['--world', world])
	args.extend(['--port', str(port)])
	return args

def try_get(priority_order, none_values=[None], default=None):
	'''
	Tries to return the value from first element to last element if it isn't a none_value, if the last case fails returns default.
	'''
	for p in priority_order:
		if not p in none_values:
			return p
	return default

def expect_type(name, key, ty):
	if not isinstance(key, ty):
		raise Exception(f"[Config] Type of {name} must be {ty}, but was: {key} ({type(key)})")

def configure(key, value):
	'''
	Ensures key-value is in the server config, then returns the value.
	'''
	global server_config
	server_config[key] = value
	return value

# WARNING. THIS PATCH ONLY WORKS ON VANILLA JARS
def log4j_patch(server_version, server_jar_location, jvm_arguments):
	global safety

	if (safety == "INSECURE-"):
		print("PROCEEDING LAUNCH (INSECURE)!")
		return
	for arg in jvm_arguments:
		if (arg.startswith("-Dlog4j2.formatMsgNoLookups") or arg.startswith("-Dlog4j.configurationFile")):
			import random
			pyprint("JVM arguments for Log4J should not be included within config.json:", 2)
			print("  A patch for the log4J [CVE-2021-44228] vulnerability will automatically be applied for any VANILLA version.")
			print("  For any version before 18w47b, the default mitigation chosen is to strip any malicious messages from logs.")
			print("  If you decide to override these choices (for example for versions prior to 1.7 or modded),")
			print("  you may still continue at the risk of your system being compromised by an outside attacker.")
			print('')
			print("  Are you sure you know what you're doing?")
			security_question = f"YES I ACKNOWLEDGE {random.randint(1000, 9999)}"
			print(f"  (To proceed, write this exact string without the quotes: '{security_question}')")
			print('')
			ans = input("  Proceed? > ")
			if (ans != security_question):
				raise Exception("Server launch has been aborted due to security violation.")
			print("PROCEEDING LAUNCH (INSECURE)!")
			safety = "INSECURE-"
			return

	# Snapshots will run according to this unless they are 22w+
	log4j_jvm_patch_flags = "-Dlog4j.configurationFile=log4j2_17-111.xml"
	patch_method = 2
	vulnerable = True
	safety = ""

	full_release_pattern = re.compile(r"^(\d+)(?:\.(\d+))?(?:\.(\d+))?(?:[_-].*)?$")
	m = full_release_pattern.match(server_version['id'])
	l = 0 if not m else len(m.groups())
	if (m and l > 2):
		major = int(m.group(1))
		minor = int(m.group(2))
		patch = int(m.group(2)) if (l > 3) else 0
		if (major != 1):
			pass
		elif ((minor == 18 and patch > 0) or (minor > 18)):
			vulnerable = False
		elif (minor == 17 or (minor == 18 and patch < 1)):
			log4j_jvm_patch_flags = "-Dlog4j2.formatMsgNoLookups=true"
			patch_method = 0
		elif (minor >= 12 and minor <= 16):
			log4j_jvm_patch_flags = "-Dlog4j.configurationFile=log4j2_112-116.xml"
			patch_method = 1
		elif (minor >= 7 and minor <= 11):
			log4j_jvm_patch_flags = "-Dlog4j.configurationFile=log4j2_17-111.xml"
			patch_method = 2
		else:
			pass
	else:
		snapshot_pattern = re.compile(r"^(\d+)(?:\.(\d+))?(?:\.(\d+))?(?:[_-].*)?$")
		m = snapshot_pattern.match(server_version['id'])
		if (m):
			year = int(m.group(1))
			week = int(m.group(2))
			patch = m.group(3)
			if (year >= 22):
				vulnerable = False
			elif (year == 21 and week >= 37):
				log4j_jvm_patch_flags = "-Dlog4j2.formatMsgNoLookups=true"
				patch_method = 0
			elif (year >= 17 and week >= 31):
				log4j_jvm_patch_flags = "-Dlog4j.configurationFile=log4j2_112-116.xml"
				patch_method = 1
			elif (year >= 13 and week >= 47):
				log4j_jvm_patch_flags = "-Dlog4j.configurationFile=log4j2_17-111.xml"
				patch_method = 2
			else:
				pass

	if (vulnerable):
		jvm_arguments.append(log4j_jvm_patch_flags)
		
		if (patch_method > 0):
			if (patch_method == 2):
				url = "https://launcher.mojang.com/v1/objects/dd2b723346a8dcd48e7f4d245f6bf09e98db9696/log4j2_17-111.xml"
				xml_path = "log4j2_17-111.xml"
			else:
				url = "https://launcher.mojang.com/v1/objects/02937d122c86ce73319ef9975b58896fc1b491d1/log4j2_112-116.xml"
				xml_path = "log4j2_112-116.xml"
			xml_destination = os.path.join(server_jar_location, xml_path)
			cache_path = os.path.join(resources_location, xml_path)
			if not path.isfile(xml_destination):
				import shutil
				if not path.isfile(cache_path):
					import requests
					pyprint("Downloading log4j patch from mojang servers...")
					chunk_size=1024
					r = requests.get(url, stream=True)
					os.makedirs(resources_location, exist_ok=True)
					with open(cache_path, 'wb') as fd:
						for chunk in r.iter_content(chunk_size=chunk_size):
							fd.write(chunk)
				pyprint("Patching log4j...")
				shutil.copy(cache_path, xml_destination)


def obtain_launch_code(config, args):
	global server_config, server_properties, server_jar, server_version

	server_name = args.server_name
	server_config = find_server_config(server_name, config['server-list'])
	server_jar_location = path.join(servers_location, server_name)
	server_jar = path.join(server_jar_location, 'server.jar')
	if not path.exists(server_jar):
		raise Exception(f"[Config] Could not find the server located at: {server_jar}")

	with zipfile.ZipFile(server_jar) as z:
		try:
			with z.open("version.json") as f:
				server_version = json.load(f)
		except:
			server_version = {
				"id": "Unknown",
				"name": "Unknown",
				"java_component": "jre-legacy"
			}
	if not 'java_component' in server_version:
		server_version['java_component'] = "jre-legacy"

	java_component = server_version['java_component']
	default_java_path = f"C:\\Program Files (x86)\\Minecraft\\runtime\\{java_component}\\windows-x64\\{java_component}\\bin\\java.exe"
	
	java_executable = try_get([config.get('java-executable')], none_values=[None, ""], default=default_java_path)
	expect_type('java-executable', java_executable, str)

	jvm_arguments = try_get([["-" + a for a in args.jvm_arguments], config.get('jvm-args')], none_values=[None, []], default=[])
	expect_type('jvm-args', jvm_arguments, list)

	# Log4j security patch
	log4j_patch(server_version, server_jar_location, jvm_arguments)
	#raise Exception("MANUAL END")

	server_properties = get_server_properties(server_jar_location)

	# Configure
	universe = configure('universe', try_get([args.universe, server_config.get('universe')], default='worlds'))
	expect_type('universe', universe, str)
	world = configure('world', try_get([args.world, server_config.get('world'), server_properties.get('level-name')], default='world'))
	expect_type('world', world, str)
	nogui = config.get('hide-gui', True)
	expect_type('hide-gui', nogui, bool)
	forceupgrade = config.get('upgrade-all-chunks-on-version-mismatch', False)
	expect_type('upgrade-all-chunks-on-version-mismatch', forceupgrade, bool)
	port = configure('port', try_get([server_config.get('port'), int(server_properties.get('server-port'))], default=25565))
	expect_type('port', port, int)
	qport = configure('query-port', try_get([int(server_properties.get('query.port'))], default=25565))
	expect_type('query-port', qport, int)
	module_data = config.get('module-data', {})
	expect_type('module-data', module_data, dict)

	# Give modules access to their module data.
	server_config['module-data'] = module_data

	# Convenience locations (stored in RAM only)
	configure('server-root', server_jar_location)
	configure('world-root', path.join(path.join(server_jar_location, universe), world))

	server_argument_list = server_args(universe, world, nogui, forceupgrade, port)
	return [java_executable] + jvm_arguments + ['-jar', 'server.jar'] + server_argument_list

def launch_server(name, launch_code):
	wd = os.getcwd()
	os.chdir('%s/%s' % (servers_location, name))
	p = subprocess.Popen(launch_code, stdout=subprocess.PIPE, stdin=subprocess.PIPE, encoding='utf-8', bufsize=1, universal_newlines=True)
	os.chdir(wd)
	return p

def add_input(input_queue):
	global read_flag
	buff = []
	while running:
		c = sys.stdin.read(1)
		buff.append(c)
		if c == '\n':
			queue_lock.acquire()
			for i in buff:
				input_queue.put(i)
			buff = []
			read_flag = True
			queue_lock.release()
			with read_condition:
				read_condition.notify()

def list_modules():
	return ', '.join([cp.name for cp in command_providers])

def import_tool():
	global command_providers, raw_imports

	for cp in command_providers:
		cp.close() # Kill any threads first.

	importlib.invalidate_caches()
	modules = glob.glob(path.join(modules_location, "*.py"))
	imports = [path.basename(f)[:-3] for f in modules if path.isfile(f) and not f.endswith('__init__.py')]

	def import_or_reload(mod):
		for ri in raw_imports:
			if ri.__name__ == mod:
				return importlib.reload(ri)
		return importlib.import_module(mod)

	raw_imports = [import_or_reload(i) for i in imports]
	command_providers = [i.get_module() for i in raw_imports]
	pyprint('Succesfully (re)loaded modules: %s' % list_modules())

def show_help():
	print('')
	pyprint('Minecraft Commands should start with a "/" (e.g. /say, /help)')
	print('\n----- Built-in Commands -----')
	print(' - help [MODULE]: Show this help, or the help page of a MODULE.')
	print(' - modules <list|reload>: List pycraft modules or reload them.')
	print(' - stop|quit|exit: Stops the server and PyCraft (identical to /stop)')
	print('\n--- PyCraft Module Commands ---')
	if len(command_providers) == 0:
		print(f' No PyCraft Server Modules were found in {modules_location}')
	for cp in command_providers:
		print(' - %s: %s' % (', '.join(cp.patterns), cp.description))

def handle_events(message):
	for k in event_triggers:
		et = event_triggers[k]
		m = et.signature.match(message)
		if m:
			et.data = message
			et.match = m
			et.event.set()
			et.event.clear()

def perform_command(cmd, stdin):
	global running
	cmd_parts = cmd.split()
	key, sub = pu.next_cmd(cmd_parts)

	# Built-in
	if (key == 'exit' or key == 'quit' or key == 'stop'):
		if len(sub) > 0:
			key2, sub2 = pu.next_cmd(sub)
			if pu.max_cmd_len(sub2, 0, pyprint): return
			if (key2 == 'FORCE'):
				try:
					stdin.write('/stop\n')
				except:
					pass
				running = False
		else:
			try:
				stdin.write('/stop\n')
			except:
				pass
		return
	if (key == 'help'):
		if len(sub) > 0:
			key2, sub2 = pu.next_cmd(sub)
			for cp in command_providers:
				if cp.matches(key2):
					cp.help(sub2)
					return
			pyprint('Unknown module/command: %s' % key2, 3)
		else:
			show_help()
		return
	if (key == 'modules'):
		key2, sub2 = pu.next_cmd(sub)
		if pu.max_cmd_len(sub2, 0, pyprint): return
		if (key2 == 'reload'): import_tool()
		elif (key2 == 'list'): pyprint('Active modules: %s' % list_modules())
		return
	# Modules
	for cp in command_providers:
		if cp.matches(key):
			try:
				cp.execute(sub, server_config, stdin, event_triggers)
			except Exception as e:
				pyprint('%s: Performing command: %s' % (e, cmd), 3)
			return
	pyprint('Unknown command: %s' % cmd, 3)

def initial_commands(stdin):
	for cmd in server_config['initialize']:
		s = cmd.strip()
		if s.startswith('/'): stdin.write('%s\n' % s[1:])
		elif len(s) > 0: perform_command(s, stdin)

def ask_server_type(config):
	while True:
		choices = []
		print(" ====== Servers loaded from config.json ====== ")
		for i in range(len(config["server-list"])):
			sn = config["server-list"][i]['name']
			choices.append(sn)
			x = f"*{i}".rjust(3, ' ')
			print(f"{x}: {sn}")
		c = input("Select the server Name or *Index to start (*q/*quit/*exit to quit): ").strip()
		if c.startswith('*'):
			rest = c[1:]
			if rest == 'q' or rest == 'quit' or rest == 'exit':
				return None
			try:
				return choices[int(rest)]
			except ValueError:
				print(f"{rest} is not a valid index selection!")
		elif c in choices:
			return c
		else:
			try:
				int(c)
				print(f"{c} is not a valid option! (Did you mean to select the index: *{c})")
			except ValueError:
				print(f"{c} is not a valid option! (case-sensitive)")

def main():
	global running

	pyprint('Version: %s' % pycraft_server_version)
	if DEBUG: pyprint('DEBUG is enabled!')

	import_tool()
	config = read_config()
	args = arguments()

	if args.server_name is None:
		x = ask_server_type(config)
		if x is None:
			return
		args.server_name = x
	
	launch_code = obtain_launch_code(config, args)
	print(f"Version: {server_version['name']}")

	launch_str = ' '.join(launch_code)
	server_name = args.server_name
	pyprint(f'Launching "{server_name}"...')
	pyprint(f'With command "{launch_str}"', 0)
	process = launch_server(args.server_name, launch_code)
	
	running = True
	def print_callback(input_queue):
		nonlocal process
		global running, read_flag

		initial_commands(process.stdin)

		while running:
			with read_condition:
				read_condition.wait()
			if read_flag:
				queue_lock.acquire()
				s = ''
				while not input_queue.empty():
					s += input_queue.get()
				if s.startswith('/'): process.stdin.write(s[1:])
				elif len(s.strip()) > 0: perform_command(s.strip(), process.stdin)
				read_flag = False
				queue_lock.release()

	input_queue = Queue()
	input_thread = Thread(target=add_input, args=(input_queue,))
	input_thread.daemon = True
	input_thread.start()
	print_thread = Thread(target=print_callback, args=(input_queue,))
	print_thread.start()

	while (process.poll() == None):
		message = process.stdout.readline().rstrip('\r\n')
		if len(message.strip()) != 0:
			print(message)
			handle_events(message)
	
	running = False

	for cp in command_providers:
		cp.close() # Kill any threads first.
	
	with read_condition:
		read_condition.notify() # Remove the condition
	print_thread.join()
	pyprint('Server has terminated succesfully!')

if __name__ == '__main__': main()